cvat/apps/engine/task.py

```diff
@@ -48,22 +50,43 @@ def rq_handler(job, exc_type, exc_value, traceback):
 ############################# Internal implementation for server API
 
 class _FrameExtractor:
-    def __init__(self, source_path, compress_quality, flip_flag=False):
+    def __init__(self, source_path, compress_quality, step=1, start=0, stop=0, flip_flag=False):
         # translate inversed range 1:95 to 2:32
         translated_quality = 96 - compress_quality
         translated_quality = round((((translated_quality - 1) * (31 - 2)) / (95 - 1)) + 2)
+        self.source = source_path
         self.output = tempfile.mkdtemp(prefix='cvat-', suffix='.data')
         target_path = os.path.join(self.output, '%d.jpg')
         output_opts = '-start_number 0 -b:v 10000k -vsync 0 -an -y -q:v ' + str(translated_quality)
+        filters = ''
+        if stop > 0:
+            filters = 'between(n,' + str(start) + ',' + str(stop) + ')'
+        elif start > 0:
+            filters = 'gte(n,' + str(start) + ')'
+        if step > 1:
+            filters += ('*' if filters else '') + 'not(mod(n,' + str(step) + '))'
+        if filters:
+            filters = "select=\"'" + filters + "'\""
         if flip_flag:
-            output_opts += ' -vf "transpose=2,transpose=2"'
+            filters += (',' if filters else '') + 'transpose=2,transpose=2'
+        if filters:
+            output_opts += ' -vf ' + filters
         ff = FFmpeg(
             inputs  = {source_path: None},
             outputs = {target_path: output_opts})
 
         slogger.glob.info("FFMpeg cmd: {} ".format(ff.cmd))
         ff.run()
 
+    def getframenumber(self):
+        probe = FFprobe(
```

Comment:
I deleted my previous comment for the code. In general I believe the code will give us incorrect value in many cases (just because it doesn't use `-vsync 0`).

Reply:
OK. I have found a way to calculate stop_frame without using FFprobe. Is stop_frame excluded from the range. I mean do we have [start, stop] or [start, stop)?

For [start, stop):
numberOfFilesInOutDir = len(glob('frame*.jpg'))
stop_frame = start_frame + numberOfFilesInOutputDir * step

In general stop_frame can be slightly incorrect. For example, if step != 1 stop_frame can be in the range [real_stop_frame - step + 1, real_stop_frame] but it isn't very important because it was not specified by the user and doesn't affect a frame calculation.

Reply:
I use FFprobe because it can get total frame number from video file without extracting all frame.  It seems get the number from metadata because it very fast.
stop_frame = total_frame_number - 1, so the frame range is [start, stop].

Reply:
Yes, the frame number of the last picture may not equal to stop_frame.
last_frame_number <= stop_frame

Reply:
The FFprobe is called when user doesn't specify stop_frame. I think save the real frame number to database is better than 0.

Reply:
@zliang7, I understand that FFprobe is called when user doesn't specify stop_frame. I just suggest don't call it at all. See how to calculate it in my comments.